### ISSUE TITLE
Training nanodet from scratch 

### ISSUE THREAD

**Issue by Cloudz333** [RangiLyu/nanodet#229]
Hi,
I'm training NanoDet-m model (ShuffleNetV2 1.0x | 320*320) from scratch with Coco dataset and 4 GeForce RTX 2080 Ti. 
Convergence seems pretty slow, it could take 1-2 weeks. 

May I ask how long did it takes for you to reach 20.6 mAP, and which setup did you use?

Thank you.

**Comment by RangiLyu** [comment]
I trained nanodet-m on Coco with one 3090 24G GPU and one i9 9900k CPU. The training process toke about 30 hours using the training configuration in config folder.

**Comment by xv1nx** [comment]
have you tried to train from scratch with multiple GPUs? I am trying to train from scratch using the default configuration and with COCO dataset, but the mAP remain very low.

![Screenshot 2021-04-23 185857](https://user-images.githubusercontent.com/1759396/115905338-44091f00-a466-11eb-80f8-1be015d1f9e0.png)
![Screenshot 2021-04-23 190039](https://user-images.githubusercontent.com/1759396/115905340-44a1b580-a466-11eb-9c16-0ae369332f70.png)

**Comment by RangiLyu** [comment]
Maybe there were some bugs after using pytorch lightning trainer. Try to use tools/deprecated/train.py to see if there is still a problem.

**Comment by Cloudz333** [comment]
I tried now to train with:
- Script: tools/deprecated/train.py
- Config File: nanodet-m.yml
- Pretrained Backbone: shufflenetv2_x1-5666bf0f80.pth
- Data: COCO train2017 (18GB) & val2017 (1GB), with annotations "2017 Train/Val annotations [241MB]" instances_train2017.json & instances_val2017.json.
- 1 GPU 

But I am still not able to reproduce the results .. The model starts from mAP 0.1 and after 100 epochs converges to 0.12 mAP.

**Comment by AlbertiPot** [comment]
> I tried now to train with:
> 
> * Script: tools/deprecated/train.py
> * Config File: nanodet-m.yml
> * Pretrained Backbone: shufflenetv2_x1-5666bf0f80.pth
> * Data: COCO train2017 (18GB) & val2017 (1GB), with annotations "2017 Train/Val annotations [241MB]" instances_train2017.json & instances_val2017.json.
> * 1 GPU
> 
> EDIT: now it converges with tools/deprecated/train.py, the problem was an inconsistency from my side.
> 
> The training with the new train.py script (multi gpu) is still not converging. As you mentioned, there might be some bugs with pytorch lightning.
> 
> Thank you!

Both single GPU and DDP mode with 2 GPUs are convergent when training with `tools/deprecated/train.py`  :
- single GPU (lr 0.14 & batch_size 192) reaches mAP 0.2047 at epoch 280

- DDP with 2 GPUs (lr 0.14 & batch_size 96 per GPU) reaches mAP 0.2034 at epoch 280

However, the distributed training with pytorch-lighting(`tools/train.py`) does not converge, mAP reporting 0.074 at epoch 190

**Comment by RangiLyu** [comment]
> > I tried now to train with:
> > 
> > * Script: tools/deprecated/train.py
> > * Config File: nanodet-m.yml
> > * Pretrained Backbone: shufflenetv2_x1-5666bf0f80.pth
> > * Data: COCO train2017 (18GB) & val2017 (1GB), with annotations "2017 Train/Val annotations [241MB]" instances_train2017.json & instances_val2017.json.
> > * 1 GPU
> > 
> > EDIT: now it converges with tools/deprecated/train.py, the problem was an inconsistency from my side.
> > The training with the new train.py script (multi gpu) is still not converging. As you mentioned, there might be some bugs with pytorch lightning.
> > Thank you!
> 
> Both single GPU and DDP mode with 2 GPUs are convergent when training with `tools/deprecated/train.py` :
> 
> * single GPU (lr 0.14 & batch_size 192) reaches mAP 0.2047 at epoch 280
> * DDP with 2 GPUs (lr 0.14 & batch_size 96 per GPU) reaches mAP 0.2034 at epoch 280
> 
> However, the distributed training with pytorch-lighting(`tools/train.py`) does not converge, mAP reporting 0.074 at epoch 190

Thanks for your work. I'm sorry that I did not test the multi-GPU training with Pytorch lightning because I only have one GPU currently. I'm not sure whether it is PyTorch lightning's bug. I'll be grateful if anyone can find out what's going wrong with the new trainer.

**Comment by RangiLyu** [comment]
> > I tried now to train with:
> > 
> > * Script: tools/deprecated/train.py
> > * Config File: nanodet-m.yml
> > * Pretrained Backbone: shufflenetv2_x1-5666bf0f80.pth
> > * Data: COCO train2017 (18GB) & val2017 (1GB), with annotations "2017 Train/Val annotations [241MB]" instances_train2017.json & instances_val2017.json.
> > * 1 GPU
> > 
> > EDIT: now it converges with tools/deprecated/train.py, the problem was an inconsistency from my side.
> > The training with the new train.py script (multi gpu) is still not converging. As you mentioned, there might be some bugs with pytorch lightning.
> > Thank you!
> 
> Both single GPU and DDP mode with 2 GPUs are convergent when training with `tools/deprecated/train.py` :
> 
> * single GPU (lr 0.14 & batch_size 192) reaches mAP 0.2047 at epoch 280
> * DDP with 2 GPUs (lr 0.14 & batch_size 96 per GPU) reaches mAP 0.2034 at epoch 280
> 
> However, the distributed training with pytorch-lighting(`tools/train.py`) does not converge, mAP reporting 0.074 at epoch 190

Could you help to use one GPU to test the checkpoint which trained with lightning's multi-GPU backend to find out if the bug is in the training process or in the evaluation process?

**Comment by RangiLyu** [comment]
I found that when using pytorch-lightning's DDP mode, the validation_epoch_end function won't gather all results across GPU. This results in the validation bug. But the training process seems to be okay. Refer to a similar issue https://github.com/PyTorchLightning/pytorch-lightning/issues/7697

**Comment by RangiLyu** [comment]
Fixed in #254

**Comment by AlbertiPot** [comment]
> > > I tried now to train with:
> > > 
> > > * Script: tools/deprecated/train.py
> > > * Config File: nanodet-m.yml
> > > * Pretrained Backbone: shufflenetv2_x1-5666bf0f80.pth
> > > * Data: COCO train2017 (18GB) & val2017 (1GB), with annotations "2017 Train/Val annotations [241MB]" instances_train2017.json & instances_val2017.json.
> > > * 1 GPU
> > > 
> > > EDIT: now it converges with tools/deprecated/train.py, the problem was an inconsistency from my side.
> > > The training with the new train.py script (multi gpu) is still not converging. As you mentioned, there might be some bugs with pytorch lightning.
> > > Thank you!
> > 
> > 
> > Both single GPU and DDP mode with 2 GPUs are convergent when training with `tools/deprecated/train.py` :
> > 
> > * single GPU (lr 0.14 & batch_size 192) reaches mAP 0.2047 at epoch 280
> > * DDP with 2 GPUs (lr 0.14 & batch_size 96 per GPU) reaches mAP 0.2034 at epoch 280
> > 
> > However, the distributed training with pytorch-lighting(`tools/train.py`) does not converge, mAP reporting 0.074 at epoch 190
> 
> Could you help to use one GPU to test the checkpoint which trained with lightning's multi-GPU backend to find out if the bug is in the training process or in the evaluation process?

Hi, sorry for the late. The validation result comes at mAP 0.14 with batch_size 96 or 192 both for single GPU, which is the same as the deprecated training settings. You are totally correct that the pytorch lightning DDP mode does have validation bugs

Thanks for your great works! Thanks for fixing the bugs!
